### ISSUE TITLE
Support ARM32 iOS devices (iPhone 4, iPhone 5)

### ISSUE THREAD

**Issue by eseidelGoogle** [flutter/flutter#2089]
@chinmaygarde  says this should be very straightforward and is just about teaching our Mac bot to build the ARM32-bit variants of our engine, uploading them to the right place, and teaching our XCode project template to use them.

EDIT: Turns out it was not straight forward.  Work is actively underway, @cbracken is driving the effort.

**Comment by chinmaygarde** [comment]
Ninja can build a non-aarch64 engine and runner if the `--ios-force-armv7` flag is specified to GN

**Comment by chinmaygarde** [comment]
The buildbot now build this target. The tools need to be wired up though.

**Comment by eseidelGoogle** [comment]
@chinmaygarde does your Apr 8 comment mean that we're publishing the artifacts but just haven't taught the `flutter` tool to use them?  FYI @danrubel

**Comment by chinmaygarde** [comment]
We got rid of the GN flags for iOS ARMv7 because no one was using them and the tools never recognized this variant. https://github.com/flutter/engine/commit/3fe246987a67486f3f38eedb0f869a215d2a4ee0

**Comment by eseidelGoogle** [comment]
If I'm reading this right: https://www.statista.com/statistics/606147/iphone-model-device-market-share-worldwide/ 32-bit arm devices account for about 17% of the worldwide usage of idevices?  If that's true, that likely bumps the priority of this bug a little at least into milestone 4 if not 3.

**Comment by chinmaygarde** [comment]
From the tooling perspective, supporting universal applications is a bit tricky. Though we can easily prepare a fat mach binary that supports armv7 as well as aarch64 (we already do this for the Flutter framework that can be used for both simulator and device builds), we also have to prepare and package the instructions snapshot for the Dart code in multiple architectures (via gen_snapshot). Then, at runtime, depending on the architecture of the device, pick the appropriate instructions snapshot and start the shell. This would also double the code side of the application and slow down the development in non-debug product modes (since we have to snapshot twice).

Is it possible to upload a separate binary to the App Store just for non-64 bit devices? That way, newer devices wont need to download an unnecessary copy of the code. Developers for whom it is critical to support iPhone 5 and below, can just prepare one extra IPA.

**Comment by eseidelGoogle** [comment]
@vlidholt noted on gitter just now that the most common feedback he's received so far is lack of support for iPhone 5.

**Comment by malkia** [comment]
I just started using flutter more, and tried uploading to my son's iPad a simple drawing app we started making, only to find out his "iPad Mini 1st gen" is armv7 (A5?). Please add support for armv7 :)

**Comment by rxlabz** [comment]
+1 :)

**Comment by eseidelGoogle** [comment]
It looks like only the iPad Air and later are 64-bit: https://en.wikipedia.org/wiki/IPad

**Comment by Hixie** [comment]
@sethladd Any opinions on this?

**Comment by atumn** [comment]
@eseidelGoogle any update on this?

**Comment by eseidelGoogle** [comment]
We recently discussed how much work this would be and decided that mostly it was tooling work and probably 2-3 weeks of work to fix and debug the various tooling workflows which would need changes to support this additional mode.  Thankfully the Dart toolchain and Flutter build system already know how to build for this configuration.

So far we just have not done the work, mostly due to lack of interest.  If anyone has marketshare data on usage of 32-bit iOS devices, I would be interested to read such.  Right now my understanding is that there is a non-zero population of 32-bit iOS users but that they're primarily using early iPads.

**Comment by malkia** [comment]
It looks like, it'll be wasted effort really, and maintenance burden. Though, thank you for looking into it Eric + team!

**Comment by rxlabz** [comment]
according to https://developer.apple.com/support/app-store/, 87% of iOS devices are using iOS 10, so there is probably at least 13% of 32bits + few % of iphone5/iphone5C/iPad4 running iOS 10

**Comment by atumn** [comment]
Thanks for the reply.
@eseidelGoogle is there a way to use the Dart toolchain and Flutter build system manually to build versions for the armv7?

**Comment by eseidelGoogle** [comment]
https://david-smith.org/iosversionstats/ lists 28% of devices using the "AudioBooks" app as being 32-bit.

**Comment by BoaNeo** [comment]
Is there *any* way for me to build for 32-bit iOS? (You seem to hint at it above, but there is no real explanation of how?)... I took a leap of faith with Flutter and now have a customer on it with an app ready to publish, but no iPhone5 support is a bit of a deal breaker.

**Comment by BoaNeo** [comment]
Anyone? This was reported 1.5 years ago and the fix was described as "straightforward" - it's a bug that rules out a significant number of iOS devices. Enough that, to me, it basically means that Flutter is just a different way to write Android apps and not something I should ever have considered using (which is a shame, I really like both the UI framework and Dart).

I know I should have given it proper due diligence, so the error is mine, but if your stance is that you will not support 32 bit iOS, i *really* think you should put that on the download page so people don't waste their time.

**Comment by atumn** [comment]
I truly support the concert from @BoaNeo 
Most of you may not have this kind of issues because you live in USA or more developed countries.
But think about the rest of the world, I believe in a lot of still developing countries 32-bit users could amount a lot more than 28% of the devices.

**Comment by BoaNeo** [comment]
@eseidelGoogle, I really could use an answer on this - I need to make a decision to either scrap Flutter and rebuild the app on another platform or tell my customer to hang in there while we wait for a patch. I don't need this integrated into an official release of the tool-chain, I just need a way to get a build done. It does not have to be easy, just possible. Considering that my alternative is to re-write the app, almost any manual build procedure, no matter how complicated, is going to be less work...

**Comment by eseidelGoogle** [comment]
I believe with https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#ios-cross-compiling-from-mac and passing --ios-cpu=arm to tools/gn it may be possible to build flutter/engine with the armv7 support, and then use the `flutter` tool with --use-local-engine (described earlier in that doc) it might be possible to make work.

However given that we haven't actually tested that flow in some time, bugs are likely.

**Comment by chinmaygarde** [comment]
I appreciate this concern and would like to do whatever I can to help. As Eric [mentioned](https://github.com/flutter/flutter/issues/2089#issuecomment-326583857) earlier in this thread, this issue was discussed a few weeks ago. It was agreed that we had all the technical pieces necessary to support `armv7` on iOS (after all, we have to support the architecture on Android already). But, a non-trivial amount of work was necessary on the tooling front to make building universal binaries for iOS seamless. Since we were unsure of the interest then, we decided there were higher priority tasks for us to work on. Given the comments on this thread, we may have to reconsider.

I am traveling at the moment and will be back next Monday. I will work on documenting a workflow for 32-bit only iOS devices then. This workflow will like include some manual steps initially. For the curios, a summary of the steps necessary to add support for the same:

## Build the universal Flutter Framework
* Build an `armv7` variant of the `Flutter` engine. This can be achieved by building the engine locally: `./flutter/tools/gn --ios --runtime-mode release --ios-cpu=arm`.
* Build an `aarch64` variant of the engine for newer devices or make the tools fetch the same from the cloud.
* Build a fat Mach `Flutter.framework` binary using the `lipo` tool with the `-create` option.

## Build the universal AOT application snapshot.
* When the `armv7` variant of the `Flutter` engine is built, an `i386` version of the AOT snapshotter is built alongside the target artifacts. Use this to generate an `armv7` AOT snapshot of the Dart application code.
  * Note: You can only use the `i386` snapshotter to generate `armv7` AOT snapshots and `x86_64` shapshotter to generate `aarch64` AOT shapshots. You **cannot** use one for the other.
* Using the `x86_64` version of the snapshotter from the `aarch64` local build or the cloud, create the `aarch64` AOT snapshot for the Dart application code.
* Build a fat Mach `App.framework` binary using the `lipo` tool with the `-create` option.

## Post Build
* The current tooling has various post processing scripts that strip the Flutter and application binaries of unneeded artifacts, those will need to be taught that `armv7` should not be stripped.

I will test this workflow and post any patches as necessary and document the steps. As you can see, most of the steps involved are around the complexity in creating and shipping the fat Mach binaries. On Android, we only ship `armv7` artifacts. But this is not feasible on iOS because such apps will not pass App Store submission guidelines (and will not even launch on iOS 11).

I no one on the team has the spare cycles to work on this by then, I will work on this as soon as I get back on Monday.

**Comment by BoaNeo** [comment]
Thanks guys, ill give it a shot and let you know if I run into any problems

**Comment by atumn** [comment]
@chinmaygarde thanks for the great explanation! Much appreciated. We will definitely try this.

**Comment by BoaNeo** [comment]
FWIW, I gave it a shot based on the information above, but my understanding of the flutter tool chain is clearly not good enough to bridge the gaps. I'm looking forward to a more elaborate how-to :)

**Comment by BoaNeo** [comment]
@chinmaygarde, is there any update on this - maybe just an ETA? I need to get back to my client with some kind of response.

@atumn, where you able to make it work?

**Comment by atumn** [comment]
@BoaNeo As a matter of time we've stopped investing our time into flutter because of this issue. So we've decided to go native for both iOS and Android. But sure we're following flutter, we're gonna hop on it when it goes beta or further.

**Comment by hariprasadiit** [comment]
Lot of developers need to support `armv7`.  We are waiting for this, before we start using flutter.

**Comment by chinmaygarde** [comment]
An update: A few patches needed to land to work around a few issues I ran into when preparing the `armv7` builds. I am documenting the instructions to build for `armv7` on this [wiki page](https://github.com/flutter/engine/wiki/iOS-Builds-Supporting-ARMv7). Will keep updating that document as I make progress. So far, I am able to build and run for `armv7`.

**Comment by BoaNeo** [comment]
@chinmaygarde, fantastic news! Thanks for putting in the effort - I understand there's still a bit of road to cover, but it sounds like the hardest part is over. I'll be following the wiki page ;)

**Comment by BoaNeo** [comment]
@chinmaygarde, any update on this?

**Comment by chinmaygarde** [comment]
I am waiting for the Dart SDK patches to land in the engine and have it rolled in to the framework. Once that is done, I will verify and update the wiki.

**Comment by BoaNeo** [comment]
Cool, thanks. Do you have any idea of a timeframe? I don't need a release-date, but are we talking days, weeks or months?

**Comment by eseidelGoogle** [comment]
@chinmaygarde has provided instructions for manually building for 32-bit iOS in our wiki, however we don't currently have plans to further integrate 32-bit iOS support in our toolchain. :(  We would like to provide such, it's just not yet come to the top of the priority list.  Pull requests to improve our iOS 32-bit integration are always welcome were someone to wish to improve things beyond @chinmaygarde's instructions.

**Comment by BoaNeo** [comment]
Maybe I misread it, but it was my understanding that the current set of instructions only go as far as creating a 32-bit executable that I can run from X-Code, not a "fat" binary I can submit to Apple, and that some component/patch is needed to get the last step working?

**Comment by caleb-allen** [comment]
We're looking at flutter as a replacement solution, and older device support is a critical issue. Have you had success @BoaNeo?

**Comment by BoaNeo** [comment]
No, there are too many if's but's and whatnots in this for me to have any faith in a successful outcome. We'll release the current app as it is so everyone else can get an update without any further delay and then start building a new app on a more mature platform that will support iPhone5 as well. Unfortunately, this will then be the app we move forward with and the flutter app will remain an (expensive) side-track.

**Comment by eseidelGoogle** [comment]
@BoaNeo I'm very sorry you've had a negative experaince with Flutter.  I wish you the best with your app development.

@caleb-allen it looks like @sannysanoff was successful at creating a 32-bit ipa with the wiki instructions:
https://github.com/flutter/flutter/issues/7888#issuecomment-337587527

It remains a race to see whether our current large clients require this feature before Apple fully deprecates 32-bit support.  We'd love to fully support 32-bit, there are just lots of competing priorities and we have to chose which to solve in which order. :(   Further pull requests to improve the situation are always welcome as we're fully open source.

Again, apologies @BoaNeo for your negative experience.

**Comment by BoaNeo** [comment]
Actually, the worst part is that I had a great experience with Flutter, it just stumbled on the finish line. 

I understand that you have priorities and as I said in the beginning, I'm quite aware that the fault is mine for not checking up on it sooner. Also, I don't think there's any doubt that 32-bit is going to go away soon, so I fully understand why that may not be a high priority going forward. It surprises me a bit that it was not a priority 1.5 years ago when it was first reported, but that's water under the bridge.

Anyway, live and learn...

**Comment by sannysanoff** [comment]
@eseidelGoogle yes, it actually worked except the scroll on Stocks app was not smooth (Ipad 4 / ios 10 / release build) and looked kinda ugly (bug with stringWidth?): Activity title had ellipsis after 3rd letter although it had full screen width to paint, and buttons ALL had wrapped text: one last letter on 2nd line, like:
CANCE
L

I'm new to flutter, and have no further comment.

**Comment by caleb-allen** [comment]
I consulted with our iOS dev and he said this issue wasn't as big as I thought it was, since Apple support is waning for those devices anyways. We're still evaluating this vs React Native.

**Comment by johanhenselmans** [comment]
I just got bitten by the same problem. I wanted to show the submitted app to a friend of mine, who has a 2012 ipad. No dice, allthough the machine run 10.3.3. According to the stats, in which I can see who picks up what, that is the case for a lot of our users. It would really help to have a proper 32/64 bit IOS toolchain in place, no matter what Apple supports. Not everyone replaces their devices every two years, especially ipads last a lot longer. The second generation iPad (2011) runs IOS 9.3.5 which is well within the reach of the IOS 8 minimum requirements for Flutter.

**Comment by johanhenselmans** [comment]
I tried the https://github.com/flutter/engine/wiki/iOS-Builds-Supporting-ARMv7 version, but I left wondering about the last steps. The issue link seems to be fixed. I do not understand where to apply the gist patch that follows the link. Is that in the original flutter tools folder (in my case /Application/Developing/flutter)? 

Also, I want a fat armv7/arm64 flutter, not a skinny armv7 flutter, are there some instruction how to accomplish that?

**Comment by johanhenselmans** [comment]
Found an iphone 4S, did the https://github.com/flutter/engine/wiki/iOS-Builds-Supporting-ARMv7 and made the changes in the last sentences in the 'normal' flutter distribution (the gist diff contains an error, in packages/flutter_tools/lib/src/commands/build_aot.dart the arm64 is replaced with armv7s, which is iphone5 only). 
I also had to change the Xcode Build Phases no 3: Run Script xcode_backend.sh, to include export FLUTTER_ENGINE to point to the armv7 version. 
And I had to change the info.plist to change the required device capabilities from arm64 to armv7s.
It then compiled, and got installed, but then I got an arm instruction error. 
Apparently an instruction is generated that is not armv7 compatible, I have no idea how to proceed.
![screen shot 2017-10-30 at 08 03 27](https://user-images.githubusercontent.com/609497/32158420-dff6161a-bd48-11e7-8589-f82a1da70ba2.png)

![screen shot 2017-10-30 at 07 50 34]
(https://user-images.githubusercontent.com/609497/32158081-7b196806-bd47-11e7-95ad-54c5719e2e5a.png)
![screen shot 2017-10-30 at 07 35 37](https://user-images.githubusercontent.com/609497/32158082-7b327864-bd47-11e7-9bd7-fb1774b719a7.png)

**Comment by mravn-google** [comment]
@chinmaygarde Can we unblock @johanhenselmans?

**Comment by bazzisoft** [comment]
I recently started using Flutter and also ran into this issue, as I'm required to support older iPad minis. I tried following the instructions https://github.com/flutter/engine/wiki/iOS-Builds-Supporting-ARMv7 and https://github.com/flutter/engine/blob/005212683c9cdc1c580b83aec70f25b4aba9872b/CONTRIBUTING.md as above but ran into an error building the engine:

    ./flutter/tools/gn --ios --ios-cpu=arm --runtime-mode=release
    ninja -C out/ios_release_arm

Output:

    ninja: Entering directory `out/ios_release_arm/'
    [765/1697] CXX clang_x64/obj/third_party/dart/runtime/bin/gen_snapshot_dart_io/file_system_watcher_android.o
    FAILED: clang_x64/obj/third_party/dart/runtime/bin/gen_snapshot_dart_io/file_system_watcher_android.o 
    ../../buildtools/mac-x64/clang/bin/clang++ -MMD -MF clang_x64/obj/third_party/dart/runtime/bin/gen_snapshot_dart_io/file_system_watcher_android.o.d -DDART_IO_SECURE_SOCKET_DISABLED -DUSE_OPENSSL=1 -DNO_TCMALLOC -DTOOLCHAIN_VERSION=81fd7fe3a7666a212d95a42541228bb1c8fb7ec3 -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D_FORTIFY_SOURCE=2 -DNDEBUG -DNVALGRIND -DDYNAMIC_ANNOTATIONS_ENABLED=0 -DTARGET_ARCH_ARM -DNDEBUG -DTARGET_OS_MACOS -DTARGET_OS_MACOS_IOS -DDART_PRECOMPILER -DPRODUCT -I../../third_party/dart/runtime -I../../third_party -I../.. -Iclang_x64/gen -I../../third_party/zlib -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.13.sdk -mmacosx-version-min=10.10  -flto -fno-strict-aliasing -fstack-protector-all -arch x86_64 -fcolor-diagnostics -Wall -Wextra -Wendif-labels -Werror -Wno-missing-field-initializers -Wno-unused-parameter -Wpartial-availability -fvisibility=hidden -stdlib=libc++ -Wheader-hygiene -Wstring-conversion -Wthread-safety -O2 -fno-ident -fdata-sections -ffunction-sections -g2 -Werror -Wall -Wextra -Wno-unused-parameter -Wnon-virtual-dtor -Wvla -Wno-conversion-null -Woverloaded-virtual -g3 -ggdb3 -fno-rtti -fno-exceptions -O3 -fvisibility-inlines-hidden -std=c++14 -fno-rtti -fno-exceptions -c ../../third_party/dart/runtime/bin/file_system_watcher_android.cc -o clang_x64/obj/third_party/dart/runtime/bin/gen_snapshot_dart_io/file_system_watcher_android.o
    In file included from ../../third_party/dart/runtime/bin/file_system_watcher_android.cc:5:
    ../../third_party/dart/runtime/platform/globals.h:331:2: error: Mismatched Host/Target architectures.
    #error Mismatched Host/Target architectures.
     ^
    1 error generated.
    [766/1697] CXX clang_x64/obj/third_party/dart/runtime/bin/gen_snapshot_dart_io/eventhandler.o
    FAILED: clang_x64/obj/third_party/dart/runtime/bin/gen_snapshot_dart_io/eventhandler.o 
    ../../buildtools/mac-x64/clang/bin/clang++ -MMD -MF clang_x64/obj/third_party/dart/runtime/bin/gen_snapshot_dart_io/eventhandler.o.d -DDART_IO_SECURE_SOCKET_DISABLED -DUSE_OPENSSL=1 -DNO_TCMALLOC -DTOOLCHAIN_VERSION=81fd7fe3a7666a212d95a42541228bb1c8fb7ec3 -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D_FORTIFY_SOURCE=2 -DNDEBUG -DNVALGRIND -DDYNAMIC_ANNOTATIONS_ENABLED=0 -DTARGET_ARCH_ARM -DNDEBUG -DTARGET_OS_MACOS -DTARGET_OS_MACOS_IOS -DDART_PRECOMPILER -DPRODUCT -I../../third_party/dart/runtime -I../../third_party -I../.. -Iclang_x64/gen -I../../third_party/zlib -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.13.sdk -mmacosx-version-min=10.10  -flto -fno-strict-aliasing -fstack-protector-all -arch x86_64 -fcolor-diagnostics -Wall -Wextra -Wendif-labels -Werror -Wno-missing-field-initializers -Wno-unused-parameter -Wpartial-availability -fvisibility=hidden -stdlib=libc++ -Wheader-hygiene -Wstring-conversion -Wthread-safety -O2 -fno-ident -fdata-sections -ffunction-sections -g2 -Werror -Wall -Wextra -Wno-unused-parameter -Wnon-virtual-dtor -Wvla -Wno-conversion-null -Woverloaded-virtual -g3 -ggdb3 -fno-rtti -fno-exceptions -O3 -fvisibility-inlines-hidden -std=c++14 -fno-rtti -fno-exceptions -c ../../third_party/dart/runtime/bin/eventhandler.cc -o clang_x64/obj/third_party/dart/runtime/bin/gen_snapshot_dart_io/eventhandler.o
    In file included from ../../third_party/dart/runtime/bin/eventhandler.cc:5:
    In file included from ../../third_party/dart/runtime/bin/eventhandler.h:8:
    In file included from ../../third_party/dart/runtime/bin/builtin.h:13:
    In file included from ../../third_party/dart/runtime/platform/assert.h:16:
    ../../third_party/dart/runtime/platform/globals.h:331:2: error: Mismatched Host/Target architectures.
    #error Mismatched Host/Target architectures.
     ^
    1 error generated.
    [767/1697] CREATE ARCHIVE obj/flutter/sky/engine/core/libsky_core.a
    ninja: build stopped: subcommand failed.

which at first glance appears to be a conflict between `-DTARGET_ARCH_ARM` and `-arch x86_64` sent as arguments to the compiler.

Is this a known issue or am I just doing something wrong?

Appreciate the help.

- Barak

**Comment by eliezedeck** [comment]
32-bit ARMv7 is indeed still important.

**Comment by hariprasadiit** [comment]
Any update on this?

We are planning to move to hybrid development. We would love to chose flutter as our hybrid app framework, but this issue has been roadblock for us. If flutter team decides not to support 32bit devices, we have to consider other alternatives.

**Comment by eliezedeck** [comment]
Flutter is very interesting, but unfortunately, there are small things that are actually having big impacts on adoption, like this 32-bit limitation and no desktop (would be awesome to be able to ship apps for desktop from the same codebase too).

At this time, we are using Qt, and compared to the performance of Flutter apps, **Qt QML apps beats Flutter** on interactivity performance to a point you can really feel Flutter to be sluggish and much slower, especially on a lower-end phones that are still very much used. I'm not sure how all these hype around Flutter is being measured, but I think it doesn't hold true on medium to lower-end phones. Unfortunately for us, that was a major turn-off.

**Comment by eseidelGoogle** [comment]
@hariprasadiit There have been no changes to this since we last updated this bug.  I agree with @eliezedeck that we should mention what platforms Flutter supports earlier on in our documentation, including possibly at time of download.  I've filed https://github.com/flutter/flutter/issues/14437 to track that.

We've been very focused on answering users requests and so far we've had comparatively few customers request this support. :(  However it's slowly bubbling up on our list of most requested items: https://github.com/flutter/flutter/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc as we fix other bugs.

@eliezedeck we have no official Desktop support planned at this time, but I have seen others build and run Flutter on desktops in the past.  Perhaps someone will take on providing good Desktop support? https://flutter.io/faq/#can-i-use-flutter-to-build-desktop-apps

@eliezedeck Thank you for sharing your performance concerns.  We still get occasional reports of hardware-specific slowness bugs on older phones, but we continue to work on improving performance across the board.   I would be *very* interested to learn which specific devices you saw sluggishness on.   Ideally in a new bug if possible? https://github.com/flutter/flutter/issues/new  That way we can get some of those exact phones and test ourselves.  Thanks!

**Comment by eseidelGoogle** [comment]
To others who are excited to see 32-bit support, please add a thumb-up to this bug, as described in https://flutter.io/support/

We look at the list of open issues sorted by thumbs up regularly and use that for scheduling work:
https://github.com/flutter/flutter/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by eliezedeck** [comment]
@eseidelGoogle Thanks for your info and for allowing other devs to know these important information beforehand.

What I have personally tested is a Huawei Mate 8, a Samsung Galaxy A6, and an old Wiko phone with 512MB RAM and some really old graphics chip. Please understand that the Huawei Mate 8 is a late-2015 phone (and I have the 4GB RAM version), not an old phone and was kind of a powerhouse when it was launched. Yet, Flutter apps will feel slower compared to native or Qt-apps on any of these phones. I was just swiping around, and it is was visible difference, using the demo app. There was also a way to show the graph of the performance, and I saw it miss the ability to deliver 60 fps very often.

Obviously, on the old Wiko phone, it was like 8 - 15 fps. Yes, other native apps on it are silky smooth, including a Qt-based Android app that I created for testing scrolling performance. I was quickly tired using the Flutter app on the Wiko phone. Of course, that was probably a phone from 2010 or something, but still, from where I live, a lot of phones are still similar to it.

But the important point is: **even on recent phones, it will be sluggish**. Of course, I'm comparing release builds here. I know that you don't, but it would be a mistake to assume everyone is going to be using the latest and greatest high-end phones.

I have not looked at the foundation codebase, but since there is an overhead into calling native codes (AFAK Skia is a native part), thus each time there is a frame draw, there is a native call, and so, more overheads. I could be wrong, but I believe this would be a bottleneck. And even on this, I'm worried about battery life. I believe it should be a work on the Dart side to reduce this overhead.

At this time, I am still keeping Flutter on my radar because, like I said before, it's a really appealing technology and an pleasant workflow.

**Comment by Hixie** [comment]
@eliezedeck Please file a separate bug for performance on specific devices, this bug report is about ARM32 support. Thanks.

**Comment by johanhenselmans** [comment]
The problem with lacking 32 bits support for IOS is that a lot of people looking for it will not give a thumbs up. They will just move on to another technology that does support 32-bit IOS, and will never be heard from again. 

So looking at the thumbs up number for triaging bug fixes/features works for people already involved. But you will miss the features that people sorely miss that are not involved as they will simply move on, to development systems that do support 32 bits IOS.

**Comment by matejthetree** [comment]
+1

**Comment by alangumer** [comment]
+1

**Comment by berwyn** [comment]
General question for the thread -- In about a month, [all support for 32bit iOS is being sunset](https://developer.apple.com/news/?id=02152018a) and it'll be impossible to submit an app that targets ARMv7 iPhones _anyway_, how important is support for legacy devices?

**Comment by BoaNeo** [comment]
You are of course perfectly well aware that if you keep asking that question while ignoring the answers you get then, eventually, you *will* get the answer you're looking for.

**Comment by berwyn** [comment]
I'm not ignoring the thread, I perfectly understand the motivation in supporting older devices. Were this about ARM32 _in general_ I would be wholeheartedly onboard! But as of next month, all new apps on the App Store _must target iOS11_, meaning that they only way to distribute a 32bit iOS app is going to be by having the user grab the source and compile it themselves! I can't imagine that this makes for a good use of the team's time vs focusing on other things for apps that can actually ship to the App Store

**Comment by BoaNeo** [comment]
Oh, I don’t disagree, not now. I just don’t see the point of asking.

**Comment by bazzisoft** [comment]
The above link regarding "sunsetting 32 bit support for iOS" specfically talks about **new** apps only. I can't imagine Apple will in one day disable all apps & updates for all older devices that don't support iOS 11. Our use case specifically is for a customer that uses a battery of iPad mini's on an Apple enterprise plan, which will also remain valid I believe,

**Comment by alangumer** [comment]
After been playing with Flutter I would love to use it to migrate a production app, however, my only concern is that Flutter does not support ARM32 devices like the iPhone 5, that's the only thing that is holding me back from using Flutter, any thoughts? thanks!

**Comment by mit-mit** [comment]
Note: when this is eventually fixed, make sure to update the FAQ: https://github.com/flutter/website/pull/856/files

**Comment by clydebarrow** [comment]
@berwyn The announcement from Apple that all apps must support iOS 11 does not mean that 32bit devices can't be targeted - just that 64 bit support is mandatory. There are plenty of older iPads out there with 32 bit cpus that are still in use and they can still be targeted by fat binaries. I'm currently using Kotlin and RoboVM to target Android and iOS and it supports 32bit devices. Flutter sounds promising and I'd be willing to tackle the hurdle of moving to a new language but that  plus the pain of telling 10-15% of my customer base they are of no importance means I'm not going to investigate Flutter any further at this point.

**Comment by korbin-w** [comment]
+1 
Hi, I have the same requirement on ARMv7 building.  Thanks.

**Comment by eseidelGoogle** [comment]
I'm told this will block customer:gold.

**Comment by eseidelGoogle** [comment]
@cbracken has volunteered to take on wiring up the last bits.

**Comment by cbracken** [comment]
* [ ] Order devices for testing / devicelab
* [ ] Update BUILD files for 32-bit engine
* [ ] Configure Goma on Mac bots
* [ ] Configure bots to build 32-bit engine
* [ ] Land flutter_tool patches to download 32-bit darwin binaries
* [ ] Land flutter tool patches for building/running 32-bit iOS apps (ideally: for debug mode, build only the variant required for testing, otherwise universal app)
* [ ] Update project templates to eliminate to arm64 default override
* [ ] Get devicelab running test suite / benchmarks on 32-bit iOS devices to avoid regressions

**Comment by Akash35pasupathi** [comment]
@ eseidelGoogle i have a doubt in Flutter ji..

now i am as android developer .. will move on flutter for develop both android and IOS... so the thing is.. is it possible to built app with using Windows or Linux OS for target IOS ... 

if it is possible, how can i run  ios emulator on windows or linux ... thank you...

**Comment by cbracken** [comment]
Alright folks, we're getting close here. As I continue to land flutter_tools refactorings so that we can provide the option to build either a universal binary that contains both armv7 and arm64 architectures, if anyone wants to try out a 32-bit only build, applying [this patch](https://github.com/cbracken/flutter/commit/0c5a96279b0b32e8dd9a83e4a0a48056cedfdd68) to flutter_tools should get you going. You'll want to apply this on master.

**Comment by cbracken** [comment]
Quick update: I've got a [WIP patch](https://github.com/cbracken/flutter/commit/6fa32e8ed1fd8643949a2e2283d8a707c1087de9) to flutter_tools that can be applied to `master` to support building armv7-only, arm64-only, and universal iOS binaries. Details are in the commit message for the patch.

The key change is that iOS build architecture is now entirely driven by Xcode's project settings, which means no more hardcoding. Unfortunately, switching architectures requires the developer to also run `pod install` from their project's `ios` subdirectory. Before I land the above patch, I'm going to land a separate patch that automatically detects changes to the project file or Podfile and re-runs `pod install` if necessary. Once that's in, I'll land this one.

In the meantime, if anyone wants to try applying the above patch, I'd love to get any feedback or hear issues with it before I land it. Once the patch is applied, you'll want to `rm bin/cache/flutter_tools.{snapshot,stamp}` from your flutter installation to force a rebuild of the `flutter` tool.

**Comment by eseidelGoogle** [comment]
FYI @mravn-google @sigurdm (who have been considering changes to our ios build structure related to add-to-app workflows).

**Comment by mravn-google** [comment]
@cbracken Some questions:
* Does this change then force users to have Cocoapods installed, even if they do not use plugins? Does it force users through the pods repo download (>1.3GB of data)? We currently have some logic to avoid that. This iOS project template has no `ios/Podfile`; we create one, if you add a plugin dependency.
* AFAIK, communicating information from Flutter tooling to Xcode has so far gone through the `Generated.xcconfig` file that Flutter tooling writes and Xcode build scripts read. Are we changing that pattern now?

Minor thing: https://github.com/flutter/flutter/pull/17210 also ensures that `pod install` is run, if the `ios/Podfile.lock` is older than `ios/Podfile` (and we have plugins). We may have a merge conflict.

**Comment by cbracken** [comment]
> Does this change then force users to have Cocoapods installed [...]?

Nope this change doesn't rely on Cocoapods at all. What it does do is remove the hardcoded `-arch arm64` from `buildXcodeProject()` when building for device, which causes us to use the architectures defined by the user in their Xcode project.

However, while I was testing that I noticed that when a project *does* use Cocoapods, we don't automatically re-run `pod install` if the build settings change in the Xcode project file (in my case the setting in question was the target architecture), with the result that the generated Pods project doesn't pick up any of those changes, but really we should have been doing this all along.

**Comment by cbracken** [comment]
As of #17312 (849676fc7ff9439d9261423973558b2f675ff5cd) on `master`, armv7-only, arm64-only, and universal iOS binaries should now be working for profile and release builds. There's one remaining VM fix required for debug builds to work on 32-bit iOS devices, which is out in [r53822](https://dart-review.googlesource.com/c/sdk/+/53822) and should roll into Flutter within the next couple of days.

For an example of how to update your Xcode project to build universal binaries, have a look at #17350, which does this for the Flutter Gallery example. I'll be updating the other examples and our `flutter create` templates shortly.

**Comment by cbracken** [comment]
As of #17358, `flutter create` generates projects that build universal iOS binaries by default. I've updated the [armv7 wiki page](https://github.com/flutter/engine/wiki/iOS-Builds-Supporting-ARMv7) to explain how to update existing projects to generate universal binaries.

**Comment by MiguelCatalan** [comment]
Should we do anything special with the plugins too? 

I pull the `marter` branch, do the changes of at #17350. But I keep getting the following error when preparing the built for release (`flutter build ios --release`)

```
 Error output from Xcode build:
↳
    ** BUILD FAILED **
    
Xcode's output:
↳
    === BUILD TARGET Runner OF PROJECT Runner WITH CONFIGURATION Release ===
    Non-fat binary *projectdirectory*/build/ios/Release-iphoneos/Runner.app/Frameworks/geolocation.framework/geolocation is not armv7. Running lipo -info:
    Non-fat file: *projectdirectory*/build/ios/Release-iphoneos/Runner.app/Frameworks/geolocation.framework/geolocation is architecture: arm64
    Command /bin/sh failed with exit code 1
Encountered error while building for device.

```

**Comment by cbracken** [comment]
@MiguelCatalan until this rolls to dev and beta branches, ensure that you're on the `master` branch, then run flutter upgrade. 

Once that's done, and you've followed the instructions on the [wiki page](https://github.com/flutter/engine/wiki/iOS-Builds-Supporting-ARMv7) I updated a few minutes ago, try re-running `flutter build ios --release`. That *should* automatically detect the build setting change and re-run `pod install` for you if you've updated master.

If you'd prefer, you can do `cd ios; pod install; cd ..` from your project directory to update the pod project architecture to `$(ARCHS_STANDARD)`.

**Comment by MiguelCatalan** [comment]
Thanks @cbracken,  great work btw.

I just did what you told, but keep getting the same error for the plugin [geolocation](https://github.com/loup-v/geolocation) when trying to build my app for release.

Did a miss anything?

**Comment by cbracken** [comment]
Thanks for checking. It's entirely possible I need to do a pass over our plugins repo and verify each of them. I've verified the gallery, which uses pods, but I'll do a pass over the rest now. Until then, one possibility is to manually edit the build settings for the geolocation plugin in the same way as documented in the wiki and see if that gets you going. I'll quick try the geolocation plugin now.

**Comment by cbracken** [comment]
@MiguelCatalan Here's what I tried:

1. Create the project
```
flutter create --ios-language=swift geoloc_test
cd geoloc_test
```
2. Edit pubspec.yaml. Add `geoloc: ^0.2.1` to the `dependencies` section.
3. Open `ios/Runner.xcworkspace`. Select the *Runner* project. In the Build Settings tab, Set the *Deployment target* to `9.0` (the `geolocation` package has this restriction).
4. Set my development team and bundle identifier in the *Signing* section of the *General* tab.
5. Run the project on on 32-bit device. `flutter run -d MyDevice --release`

I didn't add code that made use of the geolocation plugin, but I did check the contents of the built app under `build/ios/Release-iphoneos/Runner.app/Frameworks` and it contained `geolocation.framework`. Running lipo on that framework returns:
```
Architectures in the fat file: build/ios/Release-iphoneos/Runner.app/Frameworks/geolocation.framework/geolocation are: armv7 arm64
```

Note that I'm using flutter on master at `de332ec78292c2d79fdb76034328f902c9087ee9`, and cocoapods 1.5.0. You can update cocoapods by doing `gem install cocoapods` and `pod repo update`.

Can you give that a go and let me know if that work for you?

**Comment by johanhenselmans** [comment]
When I run a "flutter run --release" on a new flutter project  I am getting this warning :

You've implemented -[<UIApplicationDelegate> application:performFetchWithCompletionHandler:], but you still need to add "fetch" to the list of your supported UIBackgroundModes in your Info.plist.
You've implemented -[<UIApplicationDelegate> application:didReceiveRemoteNotification:fetchCompletionHandler:], but you still need to add "remote-notification" to the list of your supported UIBackgroundModes in your Info.plist.

i have tried to add these settings to Info.plist in the  folder ios/Runner, but no avail. 

The test app runs ok, afterwards, in the real app I am getting an unhandled exception error:

Running pod install...                                       3.0s
Starting Xcode build...                                          
 ├─Building Dart code...                             44.0s
 ├─Assembling Flutter resources...                    2.2s
 └─Compiling, linking and signing...                 18.7s
Xcode build done.                                           82.2s
Installing and launching...                                 20.6s

To repeat this help message, press "h". To quit, press "q".
[VERBOSE-2:dart_error.cc(16)] Unhandled exception:

**Comment by afocsa** [comment]
@cbracken It works for me, too. With `flutter upgrade` from `master`.
Do you know when will this be available on `beta`? Thanks.

**Comment by jeffthefate** [comment]
I ran the instructions described on https://github.com/flutter/engine/wiki/iOS-Builds-Supporting-ARMv7 but am still getting an error trying to run my application on an iPhone 5c:

`2018-05-12 12:13:58.484 ios-deploy[77178:7194887] [ !! ] Error 0xe8000087: This application does not support this device's CPU type. AMDeviceSecureInstallApplication(0, device, url, options, install_callback, 0)
Could not install build/ios/iphoneos/Runner.app on f4eb2ac4cf3cd996b4e30d6137064b834d05780e.
Try launching Xcode and selecting "Product > Run" to fix the problem:
  open ios/Runner.xcworkspace

Error running application on iPhone.
`

**Comment by MiguelCatalan** [comment]
Sorry for the delay.

@cbracken I can confirm this is working.

I had to upgrade cocoapods `gem install cocoapods` and `pod repo update`.

Also and because upgrading from older version of cocoapods had to apply [this](https://github.com/flutter/flutter/issues/16036#issuecomment-379069146).

And everything work as expected. Thanks a lot for the help.

**Comment by ypelud** [comment]
Build and debug (simulator) work well on iPhone5c but Testflight crash.
Crash is may be not the good term because the app is still alive but goes background immediately.
Tests done on the basic flutter demo app.

Did someone succeed with testFlight on iPhone 5C ?

I create an issue for that #17619.

**Comment by cbracken** [comment]
@ypelud I’ve just landed a fix for the issue you reported about distribution through test flight. Thanks for reporting.

**Comment by wurst-case** [comment]
@cbracken looks like the wiki referenced above, https://github.com/flutter/engine/wiki/iOS-Builds-Supporting-ARMv7, is gone? Or the whole wiki looks to be gone. No such page on the framework's wiki exists. I'm still getting build errors for arm32 device simulators, trying to diagnose.

**Comment by cbracken** [comment]
@wurst-case we've taken down the wiki page since Flutter builds universal binaries by default now. We still don't support building for 32-bit i386 simulators (just x86_64). 

Can you provide some details about your usecase for 32-bit simulators and why using a 64-bit simulator device doesn't work for you?

**Comment by mleonhard** [comment]
This appears to be broken again, due to Apple discontinuing support for 32-bit builds in XCode.  This wiki page has instructions for installing old build tools to get it to work:
https://github.com/flutter/flutter/wiki/Supporting-legacy-platforms

**Comment by cbracken** [comment]
@mleonhard can you run `flutter upgrade` to be sure you're using a recent version of the SDK, and if you're still hitting issues, would you mind providing details of the build failure you're seeing? 

You're correct that 32-bit i386 binaries are no longer supported on Catalina, but we eliminated the dependency on 32-bit i386 binaries in the Flutter build over the summer. 32-bit armv7 builds for old iOS versions are still supported under Xcode 11 with the iOS 13 SDK on Catalina. Here's me building the `examples/hello_world` app just now: 
```
18:16 cbracken@momiji:hello_world (master)% % sw_vers
ProductName:	Mac OS X
ProductVersion:	10.15.1
BuildVersion:	19B88

18:16 cbracken@momiji:hello_world (master)% xcodebuild -version
Xcode 11.1
Build version 11A1027

18:17 cbracken@momiji:hello_world (master)% flutter build ios --release
Downloading ios tools...                                            2.3s
Downloading ios-profile tools...                                    1.3s
Downloading ios-release tools...                                    6.4s
Running "flutter pub get" in hello_world...                         2.5s
Building io.flutter.examples.hello-world for device (ios-release)...
Warning: Missing build name (CFBundleShortVersionString).
Warning: Missing build number (CFBundleVersion).
Action Required: You must set a build name and number in the pubspec.yaml file version field before submitting to the App Store.
Signing iOS app for device deployment using developer identity: "iPhone Developer: Chris Bracken (7RANYCR6BW)"
Running Xcode build...

 ├─Building Dart code...                                    17.0s
 ├─Generating dSYM file...                                   1.4s
 ├─Stripping debug symbols...                                0.8s
 ├─Assembling Flutter resources...                           1.3s
 └─Compiling, linking and signing...                        43.7s
Xcode build done.                                           66.7s
Built /Users/cbracken/src/flutter/flutter/examples/hello_world/build/ios/iphoneos/Runner.app.

18:19 cbracken@momiji:hello_world (master)% lipo -info /Users/cbracken/src/flutter/flutter/examples/hello_world/build/ios/iphoneos/Runner.app/Runner
Architectures in the fat file: /Users/cbracken/src/flutter/flutter/examples/hello_world/build/ios/iphoneos/Runner.app/Runner are: armv7 arm64
```
